### PR TITLE
[DOCS] Adds PUT inference API docs

### PR DESCRIPTION
**Pull request by szabosteve**
This PR adds the PUT inference API docs to the DFA API docs pool.

Related issue: https://github.com/elastic/ml-team/issues/281

Available preview: http://elasticsearch_51231.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/put-inference.html

### CODE REVIEW

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+[[ml-put-inference-path-params]]
+==== {api-path-parms-title}
+
+`<model_id>`::
+(Required, string) 
+include::{docdir}/ml/ml-shared.asciidoc[tag=model-id]
+
+
+[[ml-put-inference-query-params]]
+==== {api-query-parms-title}
```
Comment (dolaru) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]: This should be `Request body` instead

Comment (szabosteve) [reply docs/reference/ml/df-analytics/apis/put-inference.asciidoc]: Thanks! Fixed.

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 2]
```diff
@@ -0,0 +1,456 @@
+[role="xpack"]
+[testenv="basic"]
+[[put-inference]]
+=== Create {infer} trained model API
+[subs="attributes"]
+++++
+<titleabbrev>Crerate {infer} trained model</titleabbrev>
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
<titleabbrev>Create {infer} trained model</titleabbrev>
```

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+[[ml-put-inference-prereq]]
+==== {api-prereq-title}
+
+If the {es} {security-features} are enabled, you must have the following 
+built-in roles and privileges:
+
+* `machine_learning_admin`
+
+
+* source index: `read`, `view_index_metadata`
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
```
There are no additional special roles. Either they must be a cluster admin, or a `machine_learning_admin`

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+* destination index: `read`, `create_index`, `manage` and `index`
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
```

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+* cluster: `monitor_ml`
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
```

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+The definition of the trained model. See <<ml-put-inference-trained-model>> for 
+details.
+
+`description`::
+(Optional, string) 
+A human-readable description of the {infer} trained model.
+
+`input`::
+(Required, object) 
+The input field names for the model definition.
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]: we should specify `input.field_names` somewhere.

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+
+`target_mean_encoding`.`field`:::
+(Required, string)
+The field name to encode.
+
+`target_mean_encoding`.`feature_name`:::
+(Required, string) 
+The name of the resulting feature.
+
+`target_mean_encoding`.`target.map`:::
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
`target_mean_encoding`.`target_map`:::
```

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+(Required, object map of string:double) 
+Object that maps the field value to the target mean value.
+
+`target_mean_encoding`.`default.value`:::
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
`target_mean_encoding`.`default_value`:::
```

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+      }
+   },
+   "target_type":"regression"
+}
+----------------------------------
+//NOTCONSOLE
+
+
+[[ml-put-inference-aggregated-output-example]]
+===== Aggregated output examples
```
Comment (benwtrent) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]: This seems out of place. 

The rest of the examples are JSON object examples. I think we should supply examples for these objects as well. 

I am not sure if the additional text below is needed. Especially given the explanation above.

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+
+* `machine_learning_admin`
+
+For more information, see <<security-privileges>> and <<built-in-roles>>.
+
+
+[[ml-put-inference-desc]]
+==== {api-description-title}
+
+Create {infer} trained model API enables you to supply a trained model that is
```
Comment (lcawl) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
The create {infer} trained model API enables you to supply a trained model that is 
```

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+(Optional, integer) 
+The index of the left child.
+
+`tree_node`.`right_child`:::
+(Optional, integer) 
+The index of the right child.
+
+`tree_node`.`default_left`:::
+(Optional, boolean) 
+Should default to the left if the feature is missing. Defaults to `true`.
```
Comment (lcawl) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
Indicates whether to default to the left when the feature is missing. Defaults to `true`.
```

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+
+`weighted_sum`.`weights`:::
+(Required, double) 
+The weights to multiply by the input values (the inference values of the trained 
+models).
+
+`weighted_mode`::
+(Optional, object) 
+This `aggregated_output` type works with regression or classification. It takes 
+a weighted vote of the input values. The the most common input value (taking the
```
Comment (lcawl) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]:
```suggestion
a weighted vote of the input values. The most common input value (taking the 
```

File: docs/reference/ml/df-analytics/apis/put-inference.asciidoc [hunk 13]
```diff
@@ -0,0 +1,494 @@
...
+(Required, double) 
+The weights to multiply by the input values (the inference values of the trained 
+models).
+
+See <<ml-put-inference-aggregated-output-example>> for more details.
+
+
+[[ml-put-inference-example]]
+==== {api-examples-title}
+
```
Comment (lcawl) [inline docs/reference/ml/df-analytics/apis/put-inference.asciidoc]: I think it's a bit unusual to have a bunch of incomplete examples in an API reference page. I think once we have a fuller set of examples (e.g. in https://www.elastic.co/guide/en/machine-learning/master/dfanalytics-examples.html), this section should just contain a single example (even if it's an abbreviated one).

**Review by benwtrent — APPROVED** [review]

**Comment by szabosteve** [comment]
**Backport commits:**
7.x: dfc9f23
7.6: ec0964f

**Pull request merged**